src/supervisor/policyAnalysis.py

```diff
@@ -0,0 +1,394 @@
+        saveCSV(root_policy_number_trend_df,"policy-root-count",True)
+        plt.close('all') 
+
+    except Exception as e:
+        print(Fore.RED+"Error in getting number of root policies: ",e)
+        print(Style.RESET_ALL)    
+    print("=============================================")
+   
+    status=True
+    return status
+
+def getReplicatedPolicy(pc,startTime, endTime, step):
+    print("Get details of the replicated policies")
+
+    try:
+        propagated_policy = pc.custom_query('count(policy_governance_info{type="propagated"})')
+
+        propagated_policy_df = MetricSnapshotDataFrame(propagated_policy)
+        propagated_policy_df["value"]=propagated_policy_df["value"].astype(float)
+        propagated_policy_df.rename(columns={"value": "NumOfPropagatedPolicies"}, inplace = True)
+        print(propagated_policy_df[['NumOfPropagatedPolicies']].to_markdown())
+
+        
+        propagated_policy_number_trend = pc.custom_query_range(
+        query='count(policy_governance_info{type="propagated"})',
+            start_time=startTime,
+            end_time=endTime,
+            step=step,
+        )
+
+        propagated_policy_number_trend_df = MetricRangeDataFrame(propagated_policy_number_trend)
+        propagated_policy_number_trend_df["value"]=propagated_policy_number_trend_df["value"].astype(float)
+        propagated_policy_number_trend_df.index= pandas.to_datetime(propagated_policy_number_trend_df.index, unit="s")
+        propagated_policy_number_trend_df.rename(columns={"value": "PropagatedtPolicyTotal"}, inplace = True)
+        propagated_policy_number_trend_df.plot(title="Total Count of propagated Policies",figsize=(30, 15))
+        plt.savefig('../../output/policy-propagated-count.png')
+        saveCSV(propagated_policy_number_trend_df,"policy-propagated-count",True)
+        plt.close('all') 
+
+    except Exception as e:
+        print(Fore.RED+"Error in getting number of propagated policies: ",e)
+        print(Style.RESET_ALL)    
+    print("=============================================")
+   
+    status=True
+    return status
+
+
+def getReplicatedPolicyControllerResponseTime(pc,startTime, endTime, step):
+    print("Get Percentage of replicated-policy reconcile times le 1 sec")
+
+    # This should have got us the 95th percentike
+    # histogram_quantile(0.95, sum(rate(controller_runtime_reconcile_time_seconds_bucket{controller="replicated-policy"}[10m])) by (le))
+    # however, if no policies are to be replicated, this metric does not change. Therefore we do not have results for this period
+    # Thus for this type of behaviour, we cannot use this technique
+    try:
+        propagated_policy = pc.custom_query('sum without(le,instance,pod) (controller_runtime_reconcile_time_seconds_bucket{controller="replicated-policy",le="1"})/sum without(le,instance,pod) (controller_runtime_reconcile_time_seconds_count{controller="replicated-policy"})')
```

Comment:
I think this might be a tricky query. As we only sum here, this is susceptible to counter restarts https://www.robustperception.io/rate-then-sum-never-sum-then-rate/.
Also probably needs to be multiplied by 100?

Reply:
Good catch on the counter restart. Let me fix that. I thought about multiplying by 100. Also thought if we do not multiply, we can rename it to fraction. But since you caught the same thing, will multiply!

Reply:
Thinking aloud a bit.

Why are we using this metric:
1. We think this is the effect of the load being applied on the system (causes).
2. We have got all the causes. 
3. Once we have got the effect, we can find out how they mathematically relate (aka regress)
4. Once we are able to establish that mathematical relationship - we can start solving a wide range of issues.

In usual case - This should have got us the 95th percentike
`histogram_quantile(0.95, sum(rate(controller_runtime_reconcile_time_seconds_bucket{controller="replicated-policy"}[10m])) by (le))`
However, if no policies are to be replicated, this metric does not change. Therefore this percentile calculation does not yield any result. BTW this would be also seen if we change the controller to deployment (ie ootb kube controller) - because the behavior logic is the same.

Workaround - 
Since we cannot run the usual stuff, we do a `sum(of-a-counter)/sum(of-another-counter)`. You are 100% right about what risk this carries. In addition to that, we have a 0 by 0 possibility as well.

@saswatamcode given all the above constraints, I see 2 possibilities - 
1. use this above metric `sum(of-a-counter)/sum(of-another-counter)`  and before we try to run regression, smoothen it out from sudden spikes or divide by zero problem. This present code is doing data collection. When we process this - we can do this.
1. get the raw counters like `controller_runtime_reconcile_time_seconds_bucket{controller="replicated-policy", le='1'}, controller_runtime_reconcile_time_seconds_bucket{controller="replicated-policy",le='0.5'} ` etc and smoothen it. And calculate the percentile. This can all be done in pandas again while processing. This will give us the appx percentile number as well - something which we do not get here. But it will involve storing data from a host of counters for each controller.

Do you see any other options? WDYT?

Reply:
So in both these options counter restarts would still be a thing. The second option would still need careful handling. How about setting a default value to instead pick up on rates?

Something like `histogram_quantile(0.95, rate(controller_runtime_reconcile_time_seconds_bucket{controller="replicated-policy",le="1"}[5m])) OR scalar(0)`. This gives you the correct spikes when the controller is actually doing work, and 0 if not, so later on you can process the pandas result based on that assumption (something is up with non-zero)?

Again, don't want to block here! We can go with one and test it out, and then change the query if we see issues arise.

Reply:
Thanks @saswatamcode . Created a separate task [ACM-12534] to fix this.